src/com/facebook/buck/jvm/java/JavaTest.java

```diff
@@ -671,21 +672,22 @@ public ExternalTestRunnerTestSpec getExternalTestRunnerSpec(
             new AbstractExecutionStep("write classpath file") {
               @Override
               public StepExecutionResult execute(ExecutionContext context) throws IOException {
-                ImmutableSet.Builder<Path> builder = ImmutableSet.builder();
+                ImmutableList.Builder<Path> builder = ImmutableList.builder();
```

Comment:
should this be an `ImmutableSortedSet` instead?

Reply:
we need it to be a `List` to make sure classpath entries from different sources are self sorted and not among themselves.

    New Ordering for Android (each is sorted individually and added to classpath entries):
    - External Libraries (from compiledTestsLibrary)
    - Dummy R.java jar (from additionalClasspathEntries)
    - Android jar (from bootClasspathEntries)
    
    
    Previously android jar could end up at the start or end of the class-path
    list since items were sorted with absolute paths where root was not the same.
    This resulted in robolectric tests randomly failing/passing in some machines.
    
    ```
    /Users/ravi/android-sdk/platforms/android-27/optional/org.apache.http.legacy.jar
    /Users/ravi/android-sdk/platforms/android-27/android.jar
    /Users/ravi/Uber/android/buck-out/gen/tool/pa/rx/lib__src_main__output/src_main.jar
    /Users/ravi/Uber/android/buck-out/gen/lib/foundation/uv/uv/lib__src_release__output/src_release.jar
    /Users/ravi/Uber/android/buck-out/gen/lib/foundation/ut/ut/lib__src_release__output/src_release.jar
    ```
    
    But still we need to keep the bootclasspath entries in the end so that android.jar
    is at the end of the classpath list. This makes sure `unit` classes comes from junit jar and not android jar.
    
    Stacktrace:
    ```
    java.lang.RuntimeException: Stub!
        at junit.runner.Version.id(Version.java:5)
        at org.powermock.modules.junit4.common.internal.impl.JUnitVersion.getJUnitVersion(JUnitVersion.java:28)
        at org.powermock.modules.junit4.common.internal.impl.JUnitVersion.isGreaterThanOrEqualTo(JUnitVersion.java:23)
        at org.powermock.modules.junit4.PowerMockRunner.getRunnerDelegateImplClass(PowerMockRunner.java:44)
        at org.powermock.modules.junit4.PowerMockRunner.<init>(PowerMockRunner.java:34)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    ```
    
    StackOverflow - https://stackoverflow.com/a/10107368/2383356

Reply:
Yes, Ravi discovered this nuance with android installation location.

You've tried merging this change early and had some failing tests, it may be possible that this android nuance was at fault.